src/esp/sensor/SensorFactory.cpp

```diff
@@ -1,22 +1,26 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
 #include "esp/sensor/SensorFactory.h"
 
 #include "esp/scene/SceneNode.h"
 #include "esp/sensor/CameraSensor.h"
+#include "esp/sensor/Sensor.h"
 
 namespace esp {
 namespace sensor {
-sensor::SensorSuite SensorFactory::createSensors(
-    scene::SceneNode& node,
-    const sensor::SensorSetup& sensorSetup) {
-  sensor::SensorSuite sensorSuite = sensor::SensorSuite();
+void SensorFactory::createSensors(scene::SceneNode& node,
+                                  const sensor::SensorSetup& sensorSetup) {
   for (const SensorSpec::ptr& spec : sensorSetup) {
     scene::SceneNode& sensorNode = node.createChild();
     // VisualSensor Setup
     if (spec->isVisualSensorSpec()) {
       if (spec->sensorSubType == SensorSubType::Orthographic ||
           spec->sensorSubType == SensorSubType::Pinhole) {
-        sensorSuite.add(CameraSensor::create(
-            sensorNode, std::dynamic_pointer_cast<CameraSensorSpec>(spec)));
+        sensor::CameraSensor::create(
+            sensorNode,
+            std::dynamic_pointer_cast<sensor::CameraSensorSpec>(spec));
       }
```

Comment:
I'm having some trouble with initializing sensors here. Currently, the sensor is initialized and a reference is added to all the relevant node and subtree sensorSuites, as expected. However, as soon as it is done initializing and the for loop moves onto the next SensorSpec, the sensor is destructed immediately. I'm not sure why this is happening because the reference to the sensor should still exist as it was added to a map, right? Why is this happening and how can I get the Sensor to persist?

Reply:
> I'm not sure why this is happening because the reference to the sensor should still exist as it was added to a map, right? Why is this happening and how can I get the Sensor to persist?

When the smart pointer gets destroyed, the dtor of Sensor class is called, which means the sensor you just added to the map gets removed (you just coded it up by yourself in this PR).